### ISSUE TITLE
Illegal return statement

### ISSUE THREAD

**Issue by wavded** [eslint/eslint#1158]
Although not valid in browser land, this is valid in node because the way the require system works.  A module can have 'return' statement to short circuit processing.  This doesn't throw an error in JSHint but does in ESLint.

```js
if (!module.parent) return
```

Output:

```
1:27  error  Illegal return statement
```

**Comment by feross** [comment]
Node is merely wrapping the code with a self-executing function, like this:

```js
(function (require, module, exports) {
  // module code here
})(require, module, exports)
```

Couldn't ESLint add a node-mode and wrap the code before passing it into esprima?

**Comment by nzakas** [comment]
No, that would affect all position-related information. Line number and ranges would be completely off.

**Comment by btmills** [comment]
It tried something similar to solve #892 (see [084aaaa](https://github.com/btmills/eslint/commit/084aaaa2d1851379645cd088cdbd14fd59eb67e2) on my fork for the code), except that it wrapped the parsed AST instead of the entire script. However, it required some hacks around locations when reporting. If someone can figure out how to wrap scripts without requiring hacks elsewhere, that might knock both of these issues out at the same time.

**Comment by nzakas** [comment]
I think the right answer is to have the parser do it. Anything done outside the parser will be a hack.

**Comment by focusaurus** [comment]
@nzakas FYI the node wrapper actually just extends your first line as [shown here](https://github.com/joyent/node/blob/885f72105bf0b18007100550f131c85ca56536e0/src/node.js#L796) so it cleverly doesn't alter the line numbers. It's just some code that comes before your first code characters on the first line. It does add an extra line to the end but in practice that seems to not cause any problems.

**Comment by lo1tuma** [comment]
It would affect the column numbers and location ranges.

**Comment by nzakas** [comment]
@focusaurus we looked at this extensively. Adding additional code causes nontrivial issues for ESLint. We aren't theorizing about this, we tried it and found the problems.

**Comment by focusaurus** [comment]
Understood. I wasn't aware/sure of that history. Just wanted to make sure we weren't struggling solely due to line numbers.

**Comment by feross** [comment]
Esprima bug filed: https://code.google.com/p/esprima/issues/detail?id=630&thanks=630&ts=1422662287

Please star it.

**Comment by nzakas** [comment]
We're not using Esprima, we use Espree (https://github.com/eslint/espree).

**Comment by feross** [comment]
Oh, sorry for the confusion.

**Comment by nzakas** [comment]
In the next version, you can use the `globalReturn` flag to allow this. I'm going to see about enabling this by default when in the Node environment.

**Comment by feross** [comment]
@nzakas excellent!

**Comment by sam-github** [comment]
I'm quite looking forward to this, return from module scope is quite useful in node.

**Comment by nzakas** [comment]
the `globalReturn` flag shipped in the last release. You can already use it.

**Comment by sam-github** [comment]
Did I do it wrong?

With https://github.com/strongloop/strong-supervisor/commit/f4438661c1772932848c8e7d40b18dbbe28a2b6c

https://github.com/strongloop/strong-supervisor/blob/try-linters/bin/sl-run.js#L26

Is still reported as `26:9  error  Illegal return statement`

**Comment by sam-github** [comment]
Thanks for the pointer, btw. I assumed it hadn't shipped because the bug was open. Which it might be... unless I'm configuring eslint wrong.

**Comment by nzakas** [comment]
You need to enable it in your config. See http://eslint.org/docs/configuring/#specifying-language-options

This bug is open to enable this option by default for the node environment.

**Comment by sam-github** [comment]
I did enable it in my config, see strongloop/strong-supervisor@f443866, or https://github.com/strongloop/strong-supervisor/blob/try-linters/.eslintrc#L8-L10

**Comment by nzakas** [comment]
What version of ESLint are you using? It's only in the very latest and I just verified that it's working.
